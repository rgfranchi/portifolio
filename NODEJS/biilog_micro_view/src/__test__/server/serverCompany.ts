import { Response } from "miragejs";
import { ICompany } from "../../Company/CompanyData";
import { ICompanySlice } from "../../interfaces/ReduxInterface";

export const companyRoutes = (actions: any) => {
  let headers = {};
  // Company.
  actions.get("company/findDropdown", (schema: any) => {
    // retornar ID E NOME.
    let dataDropdown = schema.db.companies.map(
      ({ id, nome }: ICompanySlice) => ({
        id,
        nome,
      })
    );
    if (dataDropdown.length === 0) {
      return new Response(204, headers, "");
    }
    return new Response(200, headers, dataDropdown);
  });
  actions.get("company/findAll", (schema: any) => {
    const dataBase = schema.db.companies;
    if (dataBase.length === 0) {
      return new Response(204, headers, "");
    }
    return new Response(200, headers, dataBase);
  });
  actions.get("company_erro/findAll", () => {
    return new Response(500, headers, { errors: ["FAKE Server not respond"] });
  });
  actions.delete("company/:id", (schema: any, request: any) => {
    const id = request.params.id;
    if (id === "1") {
      return new Response(203, headers, { errors: ["Fake Server Delete"] });
    }
    if (id === "9999") {
      return new Response(400, headers, {
        errors: ["Fake Server Invalid Delete"],
      });
    }
    schema.companies.find(id).destroy();
    return new Response(202, headers, { errors: ["Fake Server Delete"] });
  });
  actions.get("company/findById/:id", (schema: any, request: any) => {
    if (request.params.id === "-9999") {
      return new Response(400, headers, {
        errors: ["Fake Server Error 400"],
      });
    }
    if (request.params.id === "-1") {
      return new Response(299, headers, { errors: ["Fake Server Error"] });
    }
    const company_data = schema.companies.find(request.params.id);
    if (company_data === null) {
      return new Response(204, headers);
    }
    return new Response(200, headers, company_data.attrs);
  });
  actions.post("/company", (schema: any, request: any) => {
    const attrs = JSON.parse(request.requestBody);
    delete attrs.id;
    if (attrs.logradouro === "RETORNA_ERRO_208") {
      // resposta Valida porem não esperada.
      return new Response(208, headers, "Erro FAKE");
    }
    if (attrs.logradouro === "RETORNA_ERRO_400") {
      // resposta Valida porem não esperada.
      return new Response(400, headers, "Erro FAKE 400");
    }
    if (attrs.logradouro === "RETORNA_ERRO_4XX") {
      // resposta Valida porem não esperada.
      return new Response(499, headers, "Erro FAKE 4XX");
    }

    return new Response(201, headers, schema.companies.create(attrs));
    // return schema.companies.create(attrs);
  });
  actions.put("/company", (schema: any, request: any) => {
    let attrs = JSON.parse(request.requestBody);
    if (attrs.logradouro === "RETORNA_ERRO_4XX") {
      // resposta Valida porem não esperada.
      return new Response(499, headers, "Erro FAKE 4XX");
    }
    return new Response(200, headers, schema.companies.create(attrs));
    // return schema.companies.create(attrs);
  });
};

/**
 * Valores MOCk para teste da entidade.
 * @type {import("../../Company/CompanyData").ICompany[]}
 */

export const listCompanyMock: ICompany[] = [
  {
    id: 1,
    nome: "BIILOG",
    codPostal: "04557-030",
    pais: "Brasil",
    estado: "São Paulo",
    cidade: "São Paulo",
    bairro: "Brooklin Paulista",
    logradouro: "Rua Marquês de Cascais",
    numero: "152",
    complemento: "Sala 2",
    nomeContato: "Edmondo Gazzillo",
    email: "edmondo.gazzillo@gmail.com",
    telefone1: "1132051654",
    telefone2: "11996135016",
    cnpj: "39.282.099/0001-05",
  },
  {
    id: 2,
    nome: "WINE",
    codPostal: "29168-090",
    pais: "Brasil",
    estado: "Espirito Santo",
    cidade: "Serra",
    bairro: "Civit II",
    logradouro: "Rua Comendador Alcídes Simão Helou",
    numero: "1478",
    complemento: null,
    nomeContato: "Fábio Ferreira",
    email: "fabioferreira@wine.com.br",
    telefone1: "11968574521",
    telefone2: null,
    cnpj: "03.640.467/0001-94",
  },
  {
    id: 3,
    nome: "LinconLN",
    codPostal: "07170-350",
    pais: "Brasil",
    estado: "São Paulo",
    cidade: "Guarulhos",
    bairro: "Vila Aeroporto",
    logradouro: "Av. Papa João Paulo I",
    numero: "1818",
    complemento: null,
    nomeContato: "Vander Levorato",
    email: "vander_levorato@lincolnelectric.com.br",
    telefone1: "1135284700 ",
    telefone2: null,
    cnpj: "78.803.809/0001-49",
  },
  {
    id: 4,
    nome: "CNHi",
    codPostal: "18087-220",
    pais: "Brasil",
    estado: "São Paulo",
    cidade: "Sorocaba",
    bairro: "Éden",
    logradouro: "Avenida Jerome Case",
    numero: "1801",
    complemento: null,
    nomeContato: "Não Informado",
    email: "servicos.rental@brasif.com.br",
    telefone1: "1533341700",
    telefone2: null,
    cnpj: "80.783.926/0001-30",
  },
];

// export const listCompanyMock: ICompany[] = [
//   {
//     id: 1,
//     nome: "instituto Charles Xavier UPDATE",
//     cnpj: "01.968.344/0001-51",
//     email: "xavier@xmen.hqs",
//     nomeContato: "",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "São Paulo",
//     logradouro: "Rua UPDATE",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A1",
//   },
//   {
//     id: 2,
//     nome: "Empresa Bootstrap A2",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap2.com",
//     nomeContato: "Contato Bootstrap2",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "Piracicaba",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A2",
//   },
//   {
//     id: 4,
//     nome: "Empresa Bootstrap B1",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap.com",
//     nomeContato: "Contato Bootstrap",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "São Joaquim",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A4",
//   },
//   {
//     id: 5,
//     nome: "Empresa Bootstrap B2",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap2.com",
//     nomeContato: "Contato Bootstrap2",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "Leme",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A5",
//   },
//   {
//     id: 6,
//     nome: "Empresa Bootstrap B3",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap3.com",
//     nomeContato: "Contato Bootstrap3",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "São Roque",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A6",
//   },
//   {
//     id: 7,
//     nome: "Empresa Bootstrap C1",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap.com",
//     nomeContato: "Contato Bootstrap",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "Pirassununga",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A7",
//   },
//   {
//     id: 8,
//     nome: "Empresa Bootstrap C2",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap2.com",
//     nomeContato: "Contato Bootstrap2",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "Araras",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A8",
//   },
//   {
//     id: 9,
//     nome: "Empresa Bootstrap C3",
//     cnpj: "07.717.737/0001-79",
//     email: "empresa@bootstrap3.com",
//     nomeContato: "Contato Bootstrap3",
//     codPostal: "05625-984",
//     pais: "Brasil",
//     cidade: "São Pedro",
//     logradouro: "Rua .....",
//     bairro: "Vila Mock",
//     numero: "123",
//     complemento: "Apto A9",
//   },
// ];

// objeto deve ter mesma ordem do teste na comparação.
/**
 * @type {import("../../Company/CompanyData").ICompany}
 */
export const companyMock: ICompany = {
  id: 5,
  nome: "BIILOG MOCK",
  codPostal: "04557-030",
  pais: "Brasil",
  estado: "São Paulo",
  cidade: "São Paulo",
  bairro: "Brooklin Paulista",
  logradouro: "Rua Marquês de Cascais",
  numero: "152",
  complemento: "Sala 2",
  nomeContato: "Edmondo Gazzillo",
  email: "edmondo.gazzillo@gmail.com",
  telefone1: "(11)3205-1654",
  telefone2: "(11)3205-1654",
  cnpj: "39.282.099/0001-05",
};

let setDelay = 100;

export let loadCompanyValidValues = async (userEvent: any, screen: any) => {
  await userEvent.type(screen.getByLabelText("Nome"), companyMock.nome);
  await userEvent.type(
    screen.getByLabelText("Contato"),
    companyMock.nomeContato
  );
  await userEvent.type(screen.getByLabelText("CNPJ"), companyMock.cnpj);
  await userEvent.type(screen.getByLabelText("E-mail"), companyMock.email);
  await userEvent.type(screen.getByLabelText("CEP"), companyMock.codPostal);
  await userEvent.type(screen.getByLabelText("Cidade"), companyMock.cidade);
  await userEvent.type(screen.getByLabelText("Bairro"), companyMock.bairro);
  await userEvent.type(screen.getByLabelText("Número"), companyMock.numero);
  await userEvent.type(
    screen.getByLabelText("Telefone Principal"),
    companyMock.telefone1
  );
  await userEvent.type(
    screen.getByLabelText("Complemento"),
    companyMock.complemento,
    {
      delay: setDelay,
    }
  );
};

export let loadCompanyInvalidValues = async (userEvent: any, screen: any) => {
  setDelay = 0;
  companyMock.nome = "AB";
  companyMock.nomeContato = "CT";
  companyMock.cnpj = "42.724.607/0001-98";
  companyMock.email = "company@xxxxx";
  companyMock.codPostal = "05625-98";
  companyMock.telefone1 = "";
  await loadCompanyValidValues(userEvent, screen);
};

export const findCompanyDropdownMock = () => {
  return listCompanyMock.map(({ id, nome }) => ({
    id,
    nome,
  }));
};
