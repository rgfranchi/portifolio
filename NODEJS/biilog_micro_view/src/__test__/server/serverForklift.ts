import {
  waitForElement,
  waitForElementToBeRemoved,
  wait,
} from "@testing-library/dom";
import { Response } from "miragejs";
import { IForklift } from "../../Forklift/ForkliftData";
import {
  listModuleMock,
  moduleMock,
  loadModuleValidValues,
  loadModuleInvalid,
} from "./serverModule";
const baseURL = "forklift";

export const forkliftRoutes = (actions: any) => {
  let headers = {};
  actions.get(`${baseURL}/findAll`, (schema: any) => {
    const dataBase = schema.db.forklifts;
    if (dataBase.length === 0) {
      return new Response(204, headers, "");
    }
    return new Response(200, headers, dataBase);
  });
  actions.get(
    `${baseURL}/findByCompany/:companyId`,
    (schema: any, request: any) => {
      const company_id = request.params.companyId;
      if (company_id.toString() === "9999") {
        return new Response(404, headers, {
          errors: ["Company Id Not Found"],
        });
      }
      const dataBase = schema.db.forklifts.where(
        (forklift: IForklift) =>
          forklift.module.companyId.toString() === company_id.toString()
      );
      if (dataBase.length === 0) {
        return new Response(204, headers, "");
      }
      return new Response(200, headers, dataBase);
    }
  );

  actions.get(`${baseURL}/findById/:id`, (schema: any, request: any) => {
    if (request.params.id === "-9999") {
      return new Response(400, headers, {
        errors: ["Fake Server Error 400"],
      });
    }
    if (request.params.id === "-1") {
      return new Response(299, headers, { errors: ["Fake Server Error"] });
    }
    const data = schema.forklifts.find(request.params.id);
    if (data === null) {
      return new Response(204, headers);
    }
    return new Response(200, headers, data.attrs);
  });

  actions.post(`/${baseURL}`, (schema: any, request: any) => {
    const attrs = JSON.parse(request.requestBody);
    delete attrs.id;
    if (attrs.modelo === "RETORNA_ERRO_208") {
      return new Response(208, headers, "Erro FAKE");
    }
    if (attrs.modelo === "RETORNA_ERRO_400") {
      return new Response(400, headers, "Erro FAKE 400");
    }
    if (attrs.modelo === "RETORNA_ERRO_4XX") {
      return new Response(499, headers, "Erro FAKE 4XX");
    }
    return new Response(201, headers, schema.forklifts.create(attrs));
    // return schema.companies.create(attrs);
  });

  actions.put(`/${baseURL}`, (schema: any, request: any) => {
    const attrs = JSON.parse(request.requestBody);
    if (attrs.modelo === "RETORNA_ERRO_400") {
      // resposta Valida porem não esperada.
      return new Response(400, headers, "Erro FAKE 400");
    }
    return new Response(200, headers, schema.forklifts.create(attrs));
    // return schema.companies.create(attrs);
  });

  actions.delete(`${baseURL}/:id`, (schema: any, request: any) => {
    const id = request.params.id;
    if (id === "1") {
      return new Response(203, headers, { errors: ["Fake Server Delete"] });
    }
    if (id === "9999") {
      return new Response(400, headers, {
        errors: ["Fake Server Invalid Delete"],
      });
    }
    schema.forklifts.find(id).destroy();
    return new Response(202, headers, { errors: ["Fake Server Delete"] });
  });

  // // Company.
  // actions.get("company/findDropdown", (schema) => {
  //   // retornar ID E NOME.
  //   let dataDropdown = schema.db.companies.map(({ id, nome }) => ({
  //     id,
  //     nome,
  //   }));
  //   return new Response(200, headers, dataDropdown);
  // });
};

/**
 * Valores MOCk para teste da entidade.
 * @type {import("../../Forklift/ForkliftData").IForklift[]}
 */
export const listForkliftMock: IForklift[] = [
  {
    id: 1,
    codigo: "201 EGU_20",
    fabricante: "Still",
    modelo: "Transp Eletr 24",
    motor: "Combustão",
    anoFabricacao: "2099",
    horimetroInicial: 99999,
    hodometroInicial: 99999,
    module: listModuleMock[0],
  },
  {
    id: 2,
    codigo: "103 R20_T.Baixa",
    fabricante: "Linde",
    modelo: "R20",
    motor: "Combustão",
    anoFabricacao: "2099",
    horimetroInicial: 99999,
    hodometroInicial: 99999,
    module: listModuleMock[1],
  },
  {
    id: 3,
    codigo: "67",
    fabricante: "Hyster",
    modelo: "Matrix R1.6 2015",
    motor: "Elétrica 48v",
    anoFabricacao: "2099",
    horimetroInicial: 99999,
    hodometroInicial: 99999,
    module: listModuleMock[2],
  },
  {
    id: 4,
    codigo: "599",
    fabricante: "Hyster",
    modelo: "Fortis",
    motor: "Combustão",
    anoFabricacao: "2099",
    horimetroInicial: 99999,
    hodometroInicial: 99999,
    module: listModuleMock[3],
  },
  {
    id: 5,
    codigo: "R-07 TKP3000",
    fabricante: "Kontatec",
    modelo: "TKP3000",
    motor: "Elétrica 24v",
    anoFabricacao: "2099",
    horimetroInicial: 99999,
    hodometroInicial: 99999,
    module: listModuleMock[4],
  },
  {
    id: 6,
    codigo: "ES-18 R30XMA3",
    fabricante: "Hyster",
    modelo: "R30XMA3",
    motor: "Elétrica 36v",
    anoFabricacao: "2099",
    horimetroInicial: 99999,
    hodometroInicial: 99999,
    module: listModuleMock[5],
  },
];

// objeto deve ter mesma ordem do teste na comparação.
export const forkliftMock: IForklift = {
  id: 7,
  codigo: "ES-18 NEW",
  fabricante: "Hyster NEW",
  modelo: "R30XMA3",
  motor: "Elétrica 36v",
  anoFabricacao: "2099",
  horimetroInicial: 77777,
  hodometroInicial: 77777,
  module: moduleMock,
};

// incluído delay pois na localização do erro não funciona se estiver maior que zero
let setDelay: number = 100;

export let loadForkliftValidValues = async (
  userEvent: any,
  fireEvent: any,
  screen: any
) => {
  // Aguarda carregar toda a página com o botão Salvar.
  await wait(() => expect(screen.getByText(/Salvar/i)).toBeInTheDocument());
  await userEvent.type(screen.getByLabelText("Código"), forkliftMock.codigo, {
    delay: 500,
  });
  await userEvent.type(
    screen.getAllByLabelText("Fabricante")[0],
    forkliftMock.fabricante
  );
  await userEvent.type(
    screen.getAllByLabelText("Modelo")[0],
    forkliftMock.modelo
  );
  await userEvent.type(screen.getByLabelText("Motor"), forkliftMock.motor);
  await userEvent.type(
    screen.getByLabelText("Ano de Fabricação"),
    forkliftMock.anoFabricacao
  );

  await loadModuleValidValues(userEvent, fireEvent, screen);

  await userEvent.type(
    screen.getByLabelText("Horímetro inicial"),
    forkliftMock.horimetroInicial.toString()
  );
  await userEvent.type(
    screen.getByLabelText("Hodômetro inicial"),
    forkliftMock.hodometroInicial.toString(),
    {
      delay: setDelay,
    }
  );
};

export let loadForkliftInvalidValues = async (
  userEvent: any,
  fireEvent: any,
  screen: any
) => {
  setDelay = 0;
  forkliftMock.codigo = "";
  loadModuleInvalid();
  await loadForkliftValidValues(userEvent, fireEvent, screen);
};
