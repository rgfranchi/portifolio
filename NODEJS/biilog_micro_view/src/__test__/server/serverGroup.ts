import { Response } from "miragejs";
import {
  ISessionAccess,
  ISessionGroup,
} from "../../interfaces/SessionInterface";
const baseURL = "group";

export const groupRoutes = (actions: any) => {
  let headers = {};
  actions.get(`${baseURL}/findAll`, (schema: any) => {
    const dataBase = schema.db.groups;
    if (dataBase.length === 0) {
      return new Response(204, headers, "");
    }
    return new Response(200, headers, dataBase);
  });
  actions.get(`${baseURL}/listGroupSelect`, (schema: any) => {
    const dataBase: any = [];
    schema.db.groups.forEach((value: any) => {
      // @ts-ignore
      dataBase.push({ id: parseInt(value.id), nome: value.nome });
    });
    if (dataBase.length === 0) {
      return new Response(204, headers, "");
    }
    return new Response(200, headers, dataBase);
  });
};

const user: ISessionAccess = {
  label: "Usuário",
  page: {
    // leitura em ...Page.tsx
    create: {
      path: "/new",
      label: "Novo Usuário",
    },
    update: {
      path: "/update",
    },
    delete: {
      path: "/delete",
    },
  },
  path: "/user", // leitura em routes.tsx
};

const company: ISessionAccess = {
  path: "/company", // leitura em routes.tsx
  label: "Empresas",
  page: {
    // leitura em ...Page.tsx
    create: {
      path: "/new",
      label: "Nova Empresa",
    },
    update: {
      path: "/update",
    },
    delete: {
      path: "/delete",
    },
  },
};

const operator: ISessionAccess = {
  path: "/operator",
  label: "Operadores",
  page: {
    // leitura em ...Page.tsx
    create: {
      path: "/new",
      label: "Inserir Operador",
    },
    update: {
      path: "/update",
    },
    delete: {
      path: "/delete",
    },
  },
};

const forklift: ISessionAccess = {
  path: "/forklift",
  label: "Equipamentos",
  page: {
    // leitura em ...Page.tsx
    create: {
      path: "/new",
      label: "Inserir Equipamento",
    },
    update: {
      path: "/update",
    },
    delete: {
      path: "/delete",
    },
  },
};

const register: ISessionAccess = {
  label: "Cadastro",
  sub_item: [
    user,
    {
      path: "/machine",
      label: "Máquina",
      page: {
        create: {
          path: "/new",
          label: "Nova Máquina",
        },
        update: {
          path: "/update",
        },
        delete: {
          path: "/delete",
        },
      },
    },
  ],
};

/**
 * Valores MOCk para teste da entidade.
 */
export const listGroupMock: ISessionGroup[] = [
  {
    id: 1,
    nome: "Root",
    access: [company, operator, register, forklift],
  },
  {
    id: 2,
    nome: "Administrador",
    access: [company, operator],
  },
  {
    id: 3,
    nome: "Usuário Comum",
    access: [operator],
  },
];
