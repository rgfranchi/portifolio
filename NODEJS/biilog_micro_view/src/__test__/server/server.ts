import { createServer, RestSerializer, Model } from "miragejs";
import { userRoutes } from "./serverUser";
import { listCompanyMock, companyRoutes } from "./serverCompany";
import { listOperatorMock, operatorRoutes } from "./serverOperator";
import { listUserMock } from "./serverUser";
import { listGroupMock, groupRoutes } from "./serverGroup";
import { listForkliftMock, forkliftRoutes } from "./serverForklift";
import { listPlaceMock, placeRoutes } from "./serverPlace";

let isEmpty = false;
// variÃ¡vel para resposta vazia, impede alimentar com array original.
/**
 *
 * @param {boolean} empty
 */
export const forceEmpty = (empty: boolean) => {
  // console.info("ForceEmpty", empty);
  isEmpty = empty;
};

export const getIsEmpty = () => isEmpty;

// https://miragejs.com/docs/main-concepts/database/
// https://miragejs.com/api/classes/db-collection/
// https://miragejs.com/docs/main-concepts/route-handlers/
export const mockServer = () => {
  return createServer({
    serializers: {
      reminder: RestSerializer.extend({
        include: ["list"],
        embed: true,
      }),
    },
    models: {
      // userCurrent: Model,
      // user: Model,
      user: Model,
      company: Model,
      operator: Model,
      group: Model,
      forklift: Model,
      place: Model,
    },
    seeds(server) {
      // server.db.loadData({ userCurrent: isEmpty ? [] : userCurrentMock });
      server.db.loadData({ users: isEmpty ? [] : listUserMock });
      server.db.loadData({ groups: isEmpty ? [] : listGroupMock });
      server.db.loadData({ companies: isEmpty ? [] : listCompanyMock });
      server.db.loadData({ operators: isEmpty ? [] : listOperatorMock });
      server.db.loadData({ forklifts: isEmpty ? [] : listForkliftMock });
      server.db.loadData({ places: isEmpty ? [] : listPlaceMock });
    },
    routes() {
      this.namespace = "api/v0";
      userRoutes(this);
      companyRoutes(this);
      operatorRoutes(this);
      groupRoutes(this);
      forkliftRoutes(this);
      placeRoutes(this);
      this.passthrough("https://viacep.com.br/ws/**");
    },
  });
};
