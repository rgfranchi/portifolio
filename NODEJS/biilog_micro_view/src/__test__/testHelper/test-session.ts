import { listGroupMock } from "../server/serverGroup";
import {
  ISessionAccess,
  ISessionSlice,
} from "../../interfaces/SessionInterface";

/**
 * Todo: Utilizar interface :ISessionSlice
 */
export const objMockEmptySession: ISessionSlice = {
  userId: 0,
  company: { id: 0, nome: "" },
  group: { id: 0, nome: "", access: [] },
};

export const objMockSession: ISessionSlice = {
  userId: 3,
  company: { id: 2, nome: "WINE SESSION" },
  group: listGroupMock[0],
};

export const objMockMenuAccess: ISessionAccess[] = [
  {
    path: "/company",
    page: {
      create: {
        label: "Nova Empresa",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/operator",
    page: {
      create: {
        label: "Inserir Operador",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/user",
    page: {
      create: {
        label: "Novo Usuário",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/machine",
    page: {
      create: {
        label: "Nova Máquina",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
  {
    path: "/forklift",
    page: {
      create: {
        label: "Inserir Equipamento",
        path: "/new",
      },
      delete: {
        path: "/delete",
      },
      update: {
        path: "/update",
      },
    },
  },
];
