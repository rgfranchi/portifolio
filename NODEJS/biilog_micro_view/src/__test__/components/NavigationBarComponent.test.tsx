import React from "react";
import { createMemoryHistory } from "history";
import {
  act,
  render,
  fireEvent,
  screen,
  cleanup,
  waitForElement,
} from "@testing-library/react";

import Enzime, { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";

import { NavigationBarComponent } from "../../components/NavigationBarComponent";
import { listGroupMock } from "../server/serverGroup";

import { Provider } from "react-redux";
import store from "../../redux/store";
import ReadSession from "../../helper/SessionHelper";
import { objMockSession } from "../testHelper/test-session";
import {
  createSessionCurrentUser,
  renderProviderRouter,
} from "../testHelper/test-provider";
import { findCompanyDropdownMock } from "../server/serverCompany";
import { mockServer } from "../server/server";
import { ICompanySlice } from "../../interfaces/ReduxInterface";

let wrapper: any = null;
let history: any = null;
// const objectMock = companyMock;
// const objectEmpty = emptyCompany;
const submitMock = jest.fn();
let server: any;

beforeEach(() => {
  jest.setTimeout(20000);
  cleanup();
  server = mockServer();
  server.logging = false;
  store.dispatch({
    type: "company/companiesLoaded",
    payload: findCompanyDropdownMock(),
  });
  history = createMemoryHistory();
  createSessionCurrentUser(store);
});

afterEach(() => {
  server.shutdown();
});

Enzime.configure({ adapter: new Adapter() });

describe("NavigationBarComponent Administrator", () => {
  let listCompanies: ICompanySlice[];
  beforeEach(async () => {
    listCompanies = server.schema.db.companies;
    await setWrapperDefault(objMockSession.company, true);
    // expect(screen.getByText(/Carregando Empresas/i)).toBeInTheDocument();
    await waitForElement(
      () => screen.getAllByText(objMockSession.company.nome).length === 2
    );
  });

  test("snapshot", async () => {
    expect(wrapper.container).toMatchSnapshot();
  });

  test("Menu Elements", async () => {
    const container = wrapper.container;
    // Primeiro item do menu
    expect(
      container.getElementsByClassName("navbar-brand")[0].textContent
    ).toEqual(objMockSession.company.nome);

    // Simula click no menu dropdown
    fireEvent.click(screen.getAllByText(objMockSession.company.nome)[1]);

    // verifica valores dropdown
    listCompanies.forEach((element: ICompanySlice) => {
      expect(screen.getAllByText(element.nome)).toBeTruthy();
    });

    // valores antes de acessar company
    const keyHistoryBefore = history.location.key;
    //@ts-ignore
    expect(store.getState().session.company.nome).toEqual(
      objMockSession.company.nome
    );
    const clickCompanyName = listCompanies[3].nome;
    fireEvent.click(screen.getByText(clickCompanyName));
    // valores após de acessar company
    expect(history.location.key).toEqual(keyHistoryBefore);
    //@ts-ignore
    expect(store.getState().session.company.nome).toEqual(clickCompanyName);

    // expect(wrapper.find("Link").at(0).text()).toEqual(
    //   objMockSession.company.nome
    // );
  });
});

const setWrapperDefault = async (company: ICompanySlice, isAdm: boolean) => {
  await act(async () => {
    wrapper = await renderProviderRouter(
      render,
      <NavigationBarComponent
        // listCompanies={session.getNavbarArray()}
        company={company}
        isAdministrator={isAdm}
      />,
      history,
      store
    );
  });
};

// /**
//  * @Todo Verificar chamada de HOME e LOGOUT ainda não criado
//  */
// test("NavigationBar NavigationBar", () => {
//   const session = new ReadSession(store.getState().session);
//   const history = createMemoryHistory();
//   history.entries[0].key = "KEY01";

//   let pushSpy = jest.spyOn(history, "push"); // or 'replace', 'goBack', etc.
//   //criar teste nav bar.
//   const wrapper = mount(
//     <Router history={history}>
//       <Provider store={store}>
//         <NavigationBarComponent
//           listCompanies={session.getNavbarArray()}
//           company={{ id: "3", nome: "Exemplo Nome 3" }}
//           loadCompanies={true}
//         />
//       </Provider>
//     </Router>
//   );
//   // Valores preenchidos
//   expect(toJson(wrapper)).toMatchSnapshot();
//   // Primeiro item do menu
//   expect(wrapper.find("Link").at(0).text()).toEqual("BIILOG");

//   // Simula click no menu dropdown
//   wrapper.find("a#nav-dropdown-company").at(0).simulate("click");

//   // Empresa selecionada
//   expect(wrapper.find("a#nav-dropdown-company").at(0).text()).toEqual(
//     "Exemplo Nome 3"
//   );
//   // Verifica snapshot com opções do menu
//   expect(toJson(wrapper)).toMatchSnapshot();
//   // Clica em outra empresa realiza primera ação (onClick)
//   wrapper.find("div.dropdown-menu").childAt(1).simulate("click");
//   // Verifica se empresa foi carregada no lugar da primeira.
//   expect(wrapper.find("a#nav-dropdown-company").at(0).text()).toEqual(
//     "Exemplo Nome 5"
//   );
//   // quantidade de link disponivel.
//   expect(wrapper.find("a.nav-link").length).toEqual(5);
//   // verifica link com valor correspondente.
//   const linkTest = wrapper.find("a[href='/company']").at(0);
//   expect(linkTest.text()).toEqual("Empresas");
//   // realiza segunda uma ação. link (https://github.com/enzymejs/enzyme/issues/516#issuecomment-511330390)
//   linkTest.simulate("click", { button: 0 });
//   expect(pushSpy).toHaveBeenCalledTimes(2);
// });
