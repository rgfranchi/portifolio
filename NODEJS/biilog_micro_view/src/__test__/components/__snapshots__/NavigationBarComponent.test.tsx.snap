// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBarComponent Administrator snapshot 1`] = `
<div>
  <nav
    class="navbar navbar-expand navbar-light bg-light"
  >
    <div
      class="flex-column navbar-nav"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        WINE SESSION
      </a>
      <div
        class="nav-item"
      >
        <div
          class="dropdown nav-item"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="nav-dropdown-company"
            role="button"
          >
            WINE SESSION
          </a>
        </div>
      </div>
      <div
        class="flex-column navbar-nav"
      >
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/company"
          >
            Empresas
          </a>
        </div>
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/operator"
          >
            Operadores
          </a>
        </div>
        <div
          class="nav-item"
        >
          <div
            class="dropdown nav-item"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              id="navDropdown_id_2"
              role="button"
            >
              Cadastro
            </a>
          </div>
        </div>
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/forklift"
          >
            Equipamentos
          </a>
        </div>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rb-event-key="#"
          href="#"
          role="button"
        >
          Logout
        </a>
      </div>
    </div>
  </nav>
</div>
`;
